hivemind/p2p/p2p_daemon.py

```diff
@@ -0,0 +1,326 @@
+        self._child = None
+        self._alive = False
+        self._listen_task = None
+        self._server_stopped = asyncio.Event()
+
+    @classmethod
+    async def create(cls, *args, quic=1, tls=1, conn_manager=1, dht_mode='dht_server', force_reachability=None,
+                     nat_port_map=True, auto_nat=True, bootstrap=False, boostrap_peers=None, use_global_ipfs=False,
+                     host_port: int = None, daemon_listen_port: int = None, **kwargs):
+        if bootstrap and boostrap_peers is None and not use_global_ipfs:
+            raise AttributeError('Trying to create with bootstrap node without bootstrap nodes list. '
+                                 'It is very dangerous, because p2pd connects to global ipfs and it is very unstable. '
+                                 'If you really want this, pass use_global_ipfs=True')
+        if boostrap_peers is not None and use_global_ipfs:
+            raise AttributeError('Non empty boostrap_nodes and use_global_ipfs=True are incompatible.'
+                                 'Choose one option: your nodes list (preferable) or global ipfs (very unstable)')
+
+        self = cls()
+        p2pd_path = Path(__file__).resolve().parents[1] / P2P.P2PD_RELATIVE_PATH
+        bpeers = cls._make_bootstrap_peers(boostrap_peers)
+        dht = cls._make_dht_mode(dht_mode)
+        freachability = cls._make_force_reachability(force_reachability)
+        proc_args = self._make_process_args(
+            str(p2pd_path), *args,
+            quic=quic, tls=tls, connManager=conn_manager,
+            natPortMap=nat_port_map, autonat=auto_nat,
+            b=bootstrap, **{**bpeers, **dht, **freachability, **kwargs})
+        self._assign_daemon_ports(host_port, daemon_listen_port)
+        for try_count in range(self.NUM_RETRIES):
+            try:
+                self._initialize(proc_args)
+                await self._identify_client(P2P.RETRY_DELAY * (2 ** try_count))
+            except Exception as e:
+                logger.debug(f"Failed to initialize p2p daemon: {e}")
+                self._kill_child()
+                if try_count == P2P.NUM_RETRIES - 1:
+                    raise
+                self._assign_daemon_ports()
+                continue
+            break
+        return self
+
+    @classmethod
+    async def replicate(cls, daemon_listen_port: int, host_port: int):
+        self = cls()
+        # There is no child under control
+        # Use external already running p2pd
+        self._child = None
+        self._alive = True
+        self._assign_daemon_ports(host_port, daemon_listen_port)
+        self._client_listen_port = find_open_port()
+        self._client = p2pclient.Client(
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._daemon_listen_port}'),
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._client_listen_port}'))
+        await self._identify_client(0)
+        return self
+
+    async def wait_peers_at_least(self, peers_num, attempts=3):
```

Comment:
I think the code `await client.wait_peers_at_least(n)` may be misleading: it is similar to the call to [asyncio.wait_for](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for) and I would expect it to wait for `n` seconds (while actually it waits for `n` peers).

I'd suggest to rename the method to `wait_for_at_least_n_peers` (the line `await client.wait_for_at_least_n_peers(n)` would not be misleading).